Edited Components > Grid (#430)
vaadin/flow-and-components-documentation: documentation/components/tutorial-flow-grid.asciidoc

> === Column Keys

Author: moved

== Column Sorting

By default, this is how column sorting in the grid works:

* The first click on the column header sorts the column.
* The second click reverses the sort order.
* The third click resets the column to its unsorted state.

If multi-sorting is enabled, the user can sort by multiple columns. The first click sorts the first column. Subsequent clicks on second and more sortable column headers, add secondary and more sort criteria.

=== Defining Column Sorting

The difference between in-memory and backend sorting is key to understanding the sorting mechanism: 

* *In-memory sorting* is sorting that is applied to items fetched from the backend, before returning them to the client.

* *Backend sorting* is a list of `QuerySortOrder` objects used to implement your own fetching logic within a `DataProvider`. See <<../binding-data/tutorial-flow-data-provider#Sorting,Data Providers>> for more.

The sorting mechanism is flexible. You can configure in-memory and backend sorting together or separately. 

The sections that follow detail options you can use to setup sorting for your grid. 

==== Using a Sort Property Name 
 

> This option includes both in-memory and backend sorting. Sorting is defined at the time of column construction and uses a sort property name.

Reviewer: ```suggestion
Using sort property, you can override or customise the property or multiple properties that are used for sorting the column. This option includes both in-memory and backend sorting. Properties is defined at the time of column construction and uses a sort property name.
```

        .setHeader("Name");
----

> * When a comparator is set for a column, it is executed for all the items sent to the client.

Reviewer: ```suggestion
```
This is wrong and what I think was originally tried to say is obsolete.

> [NOTE]

Reviewer: ```suggestion
```

Author: Removed

> A comparator that performs poorly impacts the overall grid performance.

Reviewer: This is obvious so no need to have this note here at all. 
```suggestion
```

});
----

== Styling the Grid

> // I found the original content here difficult to understand and have interpreted as best I can

Reviewer: ```suggestion
```
OK, I'll pay extra attention to it then. In the future it might be best to add this comments as review comments in github, instead of actual line comments to the pull request. We probably don't want to leave these to the final documentation files.

> Styling the `Grid` component (or any Flow component) is straightforward, but requires some Web Component and shadow-DOM knowledge. Styling depends on the components position in the DOM:

Reviewer: ```suggestion
Styling the `Grid` component (or any Flow component) requires some Web Component and shadow-DOM knowledge. Styling depends on the components position in the DOM:
```

I generally don't like that we try to make things easier by saying they are easy. Let's leave this kind of expressions away.

Author: I agree.

----

=== Styling with the Theme Property

> The default Lumo theme includes different variations that you can use to style the grid. To use these, all you need to do is provide one or more variations. 

Reviewer: ```suggestion
The default Lumo theme includes different variations that you can use to style the grid. You can provide one or more variations. 
```

*Example*: Using the `addThemeNames` method to define theme variations for the grid. 

[source, java]
----
grid.addThemeNames("no-border", "no-row-borders", "row-stripes");
----

=== Styling with CSS

> You can use normal CSS styling for the content in the grid cells. While the `Grid` component itself is in the shadow DOM, the actual values (cell contents) are in the light DOM. 

Reviewer: ```suggestion
You can use normal CSS styling for the content in the grid cells. While the `Grid` component itself is in the shadow DOM, the actual values (cell contents) are in slots and therefore the light DOM. 
```

*Example*: Setting the maximum size for images in the grid. 

[source, css]
----
vaadin-grid vaadin-grid-cell-content img {
    max-height: 4em;
}
----
* `vaadin-grid-cell-content` is in the light DOM, and the selector `vaadin-grid vaadin-grid-cell-content` points to the grid's cells.

> You can also use a "styled" class to apply more advanced styles. 

Reviewer: ```suggestion
You can also use a classes to apply styles to a specific component instance. 
```

> *Example*: Using a "styled" class for centered images with rounded borders.

Reviewer: ```suggestion
*Example*: Applying rounded borders and centering images in a Grid with "styled" class name.
```

</dom-module>
----

* This sets custom styles for a `vaadin-grid` with a "styled" class. Grid's without this class remain normal.
* `theme-for="vaadin-grid"` indicates that it is overriding `vaadin-grid -components` styling. 
* `:host(.styled)` is a selector for `vaadin-grid` that has "styled" as a class. Outside the shadow DOM this is `vaadin-grid.styled`, but because the shadow DOM is boxed in its own DOM, it is selected with `:host([selector])`.

> === Styling with CSS Variables

Reviewer: I find it quite weird that this is in Grid documentation, because this has nothing to do with Grid. But let's leave it here now and I'll create an issue to relocate it to generic styling documentation, if that even exists 🤔

Author: Noted
